library/net.cpp

```diff
@@ -1100,6 +1100,8 @@ static int dnet_socket_connect(dnet_node *node, dnet_addr_socket_set &original_l
 	}
 
 	pthread_mutex_lock(&state->lock);
+
+	assert(state->succeed_count + state->failed_count <= state->total_count); // 'less' could be on timeout
```

Comment:
Does this mean application will crash without any error, without any notice, just out of nowhere?
That's no good - you should add a log at least, and better to handle this situation somehow - return an error, reset state, but not just blindly exit.